쪽지 채팅방 쪽지 내역 조회 기능 구현

hyeonnkang opened the pull request:
- Chatroom 엔티티 및 레포지토리 추가
- Custom Exception으로  MessageGetForbiddenException 및 ChatroomIdNotFoundException 추가
- MessageService 및 MessageController 에서 쪽지 리스트 조회 메소드 추가 구현

## 관련 이슈
- close #82 

## 작업 사항
- 작업사항 작성

## 참고 자료
- 참고자료 (링크 등)

## 체크리스트
- [x]  PR의 제목은 팀 내 규칙을 준수하여 알맞게 작성하였는가?
- [x]  Merge 하는 브랜치가 올바른가? (`main` branch에 실수로 PR 생성 금지)
- [x]  팀의 코딩 컨벤션을 준수하는가?
- [x]  PR과 관련없는 변경사항이 들어가지는 않았는가?
- [x]  내 코드에 대한 자기 검토가 되었는가?
- [x]  Reviewers, Assignees, Lables, Project, Milestone은 적절하게 선택하였는가?
- [x]  관련한 issue를 닫아야 하는지 점검해보고 적용했는가?

hyeonnkang committed:
feat: #82 쪽지 채팅방 쪽지 내역 조회 구현 및 테스트

- Chatroom 엔티티 및 레포지토리 추가
- Custom Exception으로  MessageGetForbiddenException 및 ChatroomIdNotFoundException 추가
- MessageService 및 MessageController 에서 쪽지 리스트 조회 메소드 추가 구현

dldmldlsy commented on src/main/java/com/umc/withme/dto/message/MessageInfoGetResponse.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Schema(description = "쪽지를 받은 사용자 정보")
+    private MemberInfoGetResponse receiver;
+
```
주고받는 사용자들의 프로필이미지, 닉네임, 아이콘 등이 필요하기 때문에 
쪽지함 조회 api merge 되면 DTO를 MemberShortInfoResponse로 바꿔야할 것같아요!

dldmldlsy replied:
화면보면 모집글 제목도 보여주고 있어서
모집글 ID랑 제목도 같이 응답데이터로 보내주면 어떨까요?? 
id는 지금 당장 필요한 건 아니지만, 추후에 제목부분을 누르면 그 모집글의 상세화면으로 이동할 수 있어야 편리할텐데, 
그럼 프론트가 바로 조회할 수 있도록 모집글 id도 같이 보내주는 게 좋을 것같다고 생각했습니다!

hyeonnkang replied:
추후 쪽지함 조회 api merge 되면 MemberShortInfoResponse로 변경하도록 주석으로 TODO 달아놓았습니다!

dldmldlsy commented on src/main/java/com/umc/withme/controller/MessageController.java:
```diff
@@ -30,8 +34,8 @@ public class MessageController {
+     * @param receiverId           쪽지 받는 회원 id(pk)
+     * @param meetId               관련 모집글 id(pk)
      * @return 쪽지생성 성공 여부
      */
```
컨트롤러에는 @Operation이 있어서 주석필요없다고 하더라구요!! 제가 쪽지 생성 api때 안지웠어서 ㅠㅠ 혹시 지워주시면 감사하겠습니다!

Wo-ogie commented on src/main/java/com/umc/withme/exception/ExceptionType.java:
```diff
@@ -56,7 +58,13 @@ public enum ExceptionType {
+    ADDRESS_NOT_FOUND_EXCEPTION(3400, "일치하는 주소를 찾을 수 없습니다.", AddressNotFoundException.class),
+
+    // Message
+    MESSAGE_GET_FORBIDDEN_EXCEPTION(4403, "해당 채팅방에 들어갈 권한이 없습니다.", MessageGetForbiddenException.class),
```
4400이 맞지 않을까요??

Wo-ogie commented on src/main/java/com/umc/withme/service/MessageService.java:
```diff
@@ -51,4 +74,32 @@ public Long createMessage(Long senderId, Long receiverId, Long meetId, MessageCr
+            throw new ChatroomIdNotFoundException(chatroomId);
+        }
+        // 사용자가 채팅방을 볼 권한이 없을 경우 예외가 발생한다.
+        else if (member != messages.get(0).getSender() && member != messages.get(0).getReceiver()) {
+            throw new MessageGetForbiddenException(chatroomId, memberId);
+        }
+
```
else if를 if로 교체하는 것은 어떤가요?
`if (messages.isEmpty()) {}`와 연관된 조건도 아닌 것 같고, 이 경우에는 if만 사용하는 것이 클린코드 측면에서 더 좋을 것 같아요

Wo-ogie commented on src/main/java/com/umc/withme/service/MessageService.java:
```diff
@@ -22,25 +29,41 @@ public class MessageService {
+        Meet meet = meetRepository.findById(meetId).orElseThrow(() -> new MeetIdNotFoundException(meetId));
 
-        Meet meet = meetRepository.findById(meetId).orElseThrow(()-> new MeetIdNotFoundException(meetId));
+        // 쪽지의 chatroom 설정
+        Long count1 = messageRepository.countBySender_IdAndReceiver_IdAndMeet_Id(senderId, receiverId, meetId);
+        Long count2 = messageRepository.countBySender_IdAndReceiver_IdAndMeet_Id(receiverId, senderId, meetId);
+        Chatroom chatroom;
+        if (count1.equals(0L) && count2.equals(0L)) { // 새로 채팅방을 생성할 경우
+            Chatroom newChatroom = new Chatroom();
+            chatroom = chatroomRepository.save(newChatroom);
+        } else if (count1.equals(0L)) {
+            chatroom = messageRepository.findTopBySender_IdAndReceiver_IdAndMeet_Id(receiverId, senderId, meetId)
+                    .getChatroom();
+        } else {
+            chatroom = messageRepository.findTopBySender_IdAndReceiver_IdAndMeet_Id(senderId, receiverId, meetId)
+                    .getChatroom();
+        }
```
로직이 잘 그려지지 않아서 제가 잘못 이해한 것일 수도 있는데, 혹시 count query를 사용하는 이유가 기존 채팅방이 존재하는지 아닌지를 확인하려는 게 맞을까요?
그런거라면 boolean type을 return해주는 `existsBy...`를 사용하는게 코드 가독성 측면에서 더 좋지 않을까 싶습니다!

hyeonnkang committed:
Merge branch 'develop' into feature/#82-message-list-in-chatroom

hyeonnkang committed:
feat: #82 코드리뷰 반영

- MessageInfoListGetResponse 에 모임 id 및 모임 모집글 제목 추가
- MessageService의 createMessage 에서 채팅방 설정 시 count by-> existsby로 변경
- 불필요한 else if 문을 if 문으로 변경

hyeonnkang committed:
Merge branch 'develop' into feature/#82-message-list-in-chatroom

Wo-ogie commented:
충돌 해결해주세요!

hyeonnkang committed:
Merge branch 'develop' into feature/#82-message-list-in-chatroom

hyeonnkang commented:
> 충돌 해결해주세요

해결했습니다 확인 부탁드려요!😊
